CQ_DISPATCH_CMD_WRITE_PAGED support in test_dispatcher and passing tests

kmabeeTT opened the pull request:
Including the test_dispatcher/infra changes for testing CQ_DISPATCH_CMD_WRITE_PAGED that are (now passing more cases) that I excluded from previous merge. 

- Dispatcher test/infra updates (mostly new functions for paged write handling) for L1, DRAM and perf testing.
- A script for bunch of interesting testcases I've tried so far crossed with DRAM/L1, for easy verif, and comparisons (since these tests are not yet in CI, will work on that later and include a few of the more interesting cases here).

Thinking of keeping as 3 different commits, but can squash if we feel strongly.

WIP (Not included here) test_prefetcher.cpp changes.

kmabeeTT committed:
#5480: FD2.0 test_dispatcher.cpp / infra updates for CQ_DISPATCH_CMD_WRITE_PAGED

 - Initial Changes
   - Just test_dispatcher.cpp for now, with support for L1/DRAM.
   - Passes everything I've thrown at it so far for WH, and DRAM for GS after followup changes
   - Add run_paged_tests.sh for running bunch of passing cases locally and 4 arbitrary perf tests with recorded BW
   - Add a debug feature to initialize DRAM banks, was very useful for debug
     but takes a few seconds so gaurded with existing -d flag for debug
     and enable for DRAM tests (currently paged DRAM write) only.

 - FD2.0 test_dispatcher/infra fixes for multiple banks per core (storage cores) in GS
   - Update worker_data struct for storing bank_id per core too to solve GS l1 paged mismatches. Updated
     common.h and test_dispatcher.cpp, test_prefetcher.cpp
   - Add get_min_required_buffer_addr() function/usage for GS paged L1 which
     have negative bank_to_l1_offset values and so our target addr needs to be
     high enough that these neg offset do not underflow. Real buffers allocate
     from top down, to work with this.

 - FD2.0 test_dispatcher refactor + new Tests + Fixes for multiple paged-write cmds
   - Padding fixes (need to apply padding after curr cmd, not before next cmd)
   - Fix for issuing multiple back to back paged write cmds, tightly pack with start_page
     Must constrain start_page to 8 bits. Use base_addr instead when start_page > num_banks
   - Add some sanity tests with multiple paged-write cmds that caught some test issues
   - Remove start_page cmd line control, doesn't make much sense anymore,
     multiple write cmds test it.

 - Cleanup/Refactor test_dispatcher tests, pull into seperate functions
   since paged-write doesn't want page-size randomized

kmabeeTT commented:
Rebased on latest and resolved write -> write_linear conflicts in common.h (no biggie).

pgkeller commented on tests/tt_metal/tt_metal/perf_microbenchmark/dispatch/test_dispatcher.cpp:
```diff
@@ -75,12 +90,20 @@ void init(int argc, char **argv) {
-        log_info(LogTest, "  -min: min transfer size bytes (default {})", MIN_XFER_SIZE_16B << 4);
         log_info(LogTest, "    -f: prefetcher fire once, use to measure dispatcher perf w/ prefetcher out of the way (default disabled)");
-        log_info(LogTest, "    -d: wrap all commands in debug commands (default disabled)");
+        log_info(LogTest, "    -d: wrap all commands in debug commands and clear DRAM (default disabled)");
```
not sure what "and clear DRAM" means

pgkeller replied:
oh, reading other code I see.  maybe "and clear DRAM to known state"

kmabeeTT replied:
thx, updated message.

kmabeeTT committed:
#5480: FD2.0 test_prefetcher.cpp updates for paged-write + paged-read e2e verif.

 - Some small refactor/cleanup of existing test, and make
   gen_paged_read_dram_test() read unique data per loop by
   incrementing start_page and base_addr.
 - Fix a couple prints in usage help
 - New functions for paged write + read test as -t 4 for now
